### PR TITLE
Cpu kernels

### PR DESCRIPTION
**Pull request by mkstoyanov**
* CPU kernels for kronmult

### CODE REVIEW

File: src/device/asgard_kronmult_cpu.cpp [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+
+// always use one thread per kron-product
+#pragma omp parallel for
+  for (int i = 0; i < num_batch; i++)
+  {
+    if constexpr (dimensions == 1)
+    {
+      T Y[n];
+      for (int j = 0; j < n; j++)
+        Y[j] = 0;
```
Comment (quantumsteve) [inline src/device/asgard_kronmult_cpu.cpp]:
```suggestion
      T Y[n]= {0};
```

File: src/device/asgard_kronmult_cpu.cpp [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+                  pX[i][n * n * j + n * l + k] * pA[3 * i][j * lda + s];
+      for (int j = 0; j < n; j++)
+        for (int l = 0; l < n; l++)
+          for (int k = 0; k < n; k++)
+            for (int s = 0; s < n; s++)
+              W[l][s][k] += Y[l][j][k] * pA[3 * i + 1][j * lda + s];
+      for (int j = 0; j < n; j++)
+        for (int l = 0; l < n; l++)
+          for (int k = 0; k < n; k++)
+            Y[j][l][k] = 0;
```
Comment (quantumsteve) [inline src/device/asgard_kronmult_cpu.cpp]:
```suggestion
      std::fill( &Y[0][0][0], &Y[0][0][0] + sizeof(Y), T{0.});)
```

File: src/device/asgard_kronmult_cpu_general.cpp [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+                               ? n4
+                               : ((dimensions == 4)
+                                      ? n3
+                                      : ((dimensions == 3)
+                                             ? n2
+                                             : ((dimensions == 2) ? n : 1))))))
+  {
+    static_assert(1 <= dimensions and dimensions <= 6);
+  }
+  T &operator()(int i, int j, int k, int l, int m, int p)
```
Comment (quantumsteve) [inline src/device/asgard_kronmult_cpu_general.cpp:37]: Which operator() depends on the value of `dimensions`?

File: src/device/asgard_kronmult_cpu_general.cpp [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+  T &operator()(int m, int p) { return serialized[m * n + p]; }
+  T &operator()(int p) { return serialized[p]; }
+  void zero()
+  {
+    for (auto &s : serialized)
+      s = 0;
+  }
+
+private:
+  int const n5, n4, n3, n2, n;
```
Comment (quantumsteve) [inline src/device/asgard_kronmult_cpu_general.cpp:74]: Do we want Tensor to be movable, copyable?

Comment (mkstoyanov) [reply src/device/asgard_kronmult_cpu_general.cpp:74]: We don't care, this is a helper method inside the kernels and not a general structure that we would use elsewhere. I just need something that is allocated on the stack and can be indexed with multiple integers.

File: src/device/asgard_kronmult_cpu_general.cpp (hunk 3 above)
Comment (mkstoyanov) [reply src/device/asgard_kronmult_cpu_general.cpp:37]: Do you want to put a static assert in each operator? It's kind of obvious that 6 indexes mean 6D.

File: src/device/asgard_kronmult_cpu.cpp (hunk 2 above)
Comment (mkstoyanov) [reply src/device/asgard_kronmult_cpu.cpp]: Actually `sizeof(Y) / sizeof(T)` but a much better idea than the loop.

File: src/device/asgard_kronmult_cpu_general.cpp (hunk 3 above)
Comment (mkstoyanov) [reply src/device/asgard_kronmult_cpu_general.cpp:37]: Put in the static asserts.

File: src/device/asgard_kronmult_cpu_general.cpp (hunk 4 above)
Comment (mkstoyanov) [reply src/device/asgard_kronmult_cpu_general.cpp:74]: Sorry, allocatable on the heap.

This struct is not for a general use, that's why it's defined in a .cpp file.

**Review by quantumsteve — APPROVED** [review]

**Pull request merged**